--- a/lib/coreBundle.js
+++ b/lib/coreBundle.js
@@ -22956,7 +22956,7 @@
       Close: "close"
     };
     Worker = class _Worker extends SdkObject {
-      constructor(parent, url3, onDisconnect) {
+      constructor(parent, url3, onDisconnect, targetId, session2) {
         super(parent, "worker");
         this._executionContextPromise = new ManualPromise();
         this._workerScriptLoaded = false;
@@ -22965,10 +22965,22 @@
         this.attribution.worker = this;
         this.url = url3;
         this._onDisconnect = onDisconnect;
+        this._targetId = targetId;
+        this._session = session2;
       }
       static {
         this.Events = WorkerEvent;
       }
+      // rebrowser-patches: manually acquire the worker main-world context when Runtime.enable is skipped
+      async getExecutionContext() {
+        if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] !== "0" && !this.existingExecutionContext && this._session) {
+          await this._session.__re__emitExecutionContext({
+            world: "main",
+            targetId: this._targetId
+          });
+        }
+        return this._executionContextPromise;
+      }
       createExecutionContext(delegate) {
         this.existingExecutionContext = new ExecutionContext(this, delegate, "worker");
         if (this._workerScriptLoaded)
@@ -22994,10 +23006,10 @@
         this.openScope.close(new Error("Worker closed"));
       }
       async evaluateExpression(progress2, expression2, isFunction2, arg) {
-        return progress2.race(evaluateExpression(await this._executionContextPromise, expression2, { returnByValue: true, isFunction: isFunction2 }, arg));
+        return progress2.race(evaluateExpression(await this.getExecutionContext(), expression2, { returnByValue: true, isFunction: isFunction2 }, arg));
       }
       async evaluateExpressionHandle(progress2, expression2, isFunction2, arg) {
-        return progress2.race(evaluateExpression(await this._executionContextPromise, expression2, { returnByValue: false, isFunction: isFunction2 }, arg));
+        return progress2.race(evaluateExpression(await this.getExecutionContext(), expression2, { returnByValue: false, isFunction: isFunction2 }, arg));
       }
       async disconnect(progress2, options = {}) {
         if (!this._onDisconnect)
@@ -23029,6 +23041,9 @@
         this.cleanupScript = `globalThis['${kBindingsControllerProperty}'].removeBinding(${JSON.stringify(name)})`;
       }
       static async dispatch(page, payload, context) {
+        if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] !== "0" && !payload.includes("{")) {
+          return;
+        }
         const { name, seq, serializedArgs } = JSON.parse(payload);
         try {
           assert(context.world);
@@ -23505,6 +23520,8 @@
           this._startNetworkIdleTimer();
         this._page.mainFrame()._recalculateNetworkIdle(this);
         this.onLifecycleEvent("commit");
+        const crSession = (this._page.delegate._sessions?.get(this._id) || this._page.delegate._mainFrameSession)?._client;
+        crSession?.emit("Runtime.executionContextsCleared");
       }
       _setPendingDocument(documentInfo) {
         this._pendingDocument = documentInfo;
@@ -23677,13 +23694,30 @@
       async frameElement(progress2) {
         return await progress2.race(this._page.delegate.getFrameElement(this));
       }
-      context(world) {
+      context(world, useContextPromise = false) {
         if (this._page.delegate.noUtilityWorld?.())
           world = "main";
-        return this._contextData.get(world).contextPromise.then((contextOrDestroyedReason) => {
-          if (contextOrDestroyedReason instanceof ExecutionContext)
-            return contextOrDestroyedReason;
-          throw new Error(contextOrDestroyedReason.destroyedReason);
+        if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] === "0" || this._contextData.get(world).context || useContextPromise) {
+          return this._contextData.get(world).contextPromise.then((contextOrDestroyedReason) => {
+            if (contextOrDestroyedReason instanceof ExecutionContext)
+              return contextOrDestroyedReason;
+            throw new Error(contextOrDestroyedReason.destroyedReason);
+          });
+        }
+        const crSession = (this._page.delegate._sessions?.get(this._id) || this._page.delegate._mainFrameSession)._client;
+        return crSession.__re__emitExecutionContext({
+          world,
+          targetId: this._id,
+          frame: this,
+          playwrightUtilityWorldName: this._page.delegate.utilityWorldName
+        }).then(() => {
+          return this.context(world, true);
+        }).catch((error) => {
+          if (error.message.includes("No frame for given id found")) {
+            return { destroyedReason: "Frame was detached" };
+          }
+          console.error("[rebrowser-patches][frames.context] cannot get world, error:", error);
+          throw error;
         });
       }
       mainContext() {
@@ -35042,6 +35076,130 @@
         }
         this._callbacks.clear();
       }
+      async __re__emitExecutionContext({
+        world,
+        targetId,
+        frame = null,
+        playwrightUtilityWorldName = void 0
+      }) {
+        const fixMode = process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] || "addBinding";
+        const envUtilityWorldName = process.env["REBROWSER_PATCHES_UTILITY_WORLD_NAME"];
+        const playwrightUtilityName = playwrightUtilityWorldName || "__playwright_utility_world__";
+        const utilityWorldName = envUtilityWorldName !== "0" ? envUtilityWorldName || "util" : playwrightUtilityName;
+        process.env["REBROWSER_PATCHES_DEBUG"] && console.log(`[rebrowser-patches][crSession] targetId = ${targetId}, world = ${world}, frame = ${frame ? "Y" : "N"}, fixMode = ${fixMode}`);
+        let getWorldPromise;
+        if (fixMode === "addBinding") {
+          if (world === "utility") {
+            getWorldPromise = this.__re__getIsolatedWorld({
+              client: this,
+              frameId: targetId,
+              worldName: utilityWorldName
+            }).then((contextId4) => {
+              return {
+                id: contextId4,
+                // use crPage.utilityWorldName otherwise _onExecutionContextCreated will ignore it
+                name: playwrightUtilityName,
+                auxData: {
+                  frameId: targetId,
+                  isDefault: false
+                }
+              };
+            });
+          } else if (world === "main") {
+            getWorldPromise = this.__re__getMainWorld({
+              client: this,
+              frameId: targetId,
+              isWorker: frame === null
+            }).then((contextId4) => {
+              return {
+                id: contextId4,
+                name: "",
+                auxData: {
+                  frameId: targetId,
+                  isDefault: true
+                }
+              };
+            });
+          }
+        } else if (fixMode === "alwaysIsolated") {
+          getWorldPromise = this.__re__getIsolatedWorld({
+            client: this,
+            frameId: targetId,
+            worldName: utilityWorldName
+          }).then((contextId4) => {
+            return {
+              id: contextId4,
+              name: "",
+              auxData: {
+                frameId: targetId,
+                isDefault: true
+              }
+            };
+          });
+        }
+        const contextPayload = await getWorldPromise;
+        this.emit("Runtime.executionContextCreated", {
+          context: contextPayload
+        });
+      }
+      async __re__getMainWorld({ client, frameId, isWorker = false }) {
+        let contextId4;
+        const randomName = [...Array(Math.floor(Math.random() * (10 + 1)) + 10)].map(() => Math.random().toString(36)[2]).join("");
+        process.env["REBROWSER_PATCHES_DEBUG"] && console.log(`[rebrowser-patches][getMainWorld] binding name = ${randomName}`);
+        await client.send("Runtime.addBinding", {
+          name: randomName
+        });
+        const bindingCalledHandler = ({ name, payload, executionContextId }) => {
+          process.env["REBROWSER_PATCHES_DEBUG"] && console.log("[rebrowser-patches][bindingCalledHandler]", {
+            name,
+            payload,
+            executionContextId
+          });
+          if (contextId4 > 0) {
+            return;
+          }
+          if (name !== randomName) {
+            return;
+          }
+          if (payload !== frameId) {
+            return;
+          }
+          contextId4 = executionContextId;
+          client.off("Runtime.bindingCalled", bindingCalledHandler);
+        };
+        client.on("Runtime.bindingCalled", bindingCalledHandler);
+        if (isWorker) {
+          await client.send("Runtime.evaluate", {
+            expression: `this['${randomName}']('${frameId}')`
+          });
+        } else {
+          await client.send("Page.addScriptToEvaluateOnNewDocument", {
+            source: `document.addEventListener('${randomName}', (e) => self['${randomName}'](e.detail.frameId))`,
+            runImmediately: true
+          });
+          const createIsolatedWorldResult = await client.send("Page.createIsolatedWorld", {
+            frameId,
+            // use randomName for worldName to distinguish from normal utility world
+            worldName: randomName,
+            grantUniveralAccess: true
+          });
+          await client.send("Runtime.evaluate", {
+            expression: `document.dispatchEvent(new CustomEvent('${randomName}', { detail: { frameId: '${frameId}' } }))`,
+            contextId: createIsolatedWorldResult.executionContextId
+          });
+        }
+        process.env["REBROWSER_PATCHES_DEBUG"] && console.log(`[rebrowser-patches][getMainWorld] result:`, { contextId: contextId4 });
+        return contextId4;
+      }
+      async __re__getIsolatedWorld({ client, frameId, worldName }) {
+        const createIsolatedWorldResult = await client.send("Page.createIsolatedWorld", {
+          frameId,
+          worldName,
+          grantUniveralAccess: true
+        });
+        process.env["REBROWSER_PATCHES_DEBUG"] && console.log(`[rebrowser-patches][getIsolatedWorld] result:`, createIsolatedWorldResult);
+        return createIsolatedWorldResult.executionContextId;
+      }
     };
     CDPSession = class _CDPSession extends SdkObject {
       constructor(parentSession, sessionId) {
@@ -37622,7 +37780,11 @@
           }),
           this._client.send("Log.enable", {}),
           lifecycleEventsEnabled = this._client.send("Page.setLifecycleEventsEnabled", { enabled: true }),
-          this._client.send("Runtime.enable", {}),
+          (() => {
+            if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] === "0")
+              return this._client.send("Runtime.enable", {});
+            return void 0;
+          })(),
           this._client.send("Page.addScriptToEvaluateOnNewDocument", {
             source: "",
             worldName: this._crPage.utilityWorldName
@@ -37824,7 +37986,7 @@
           return;
         }
         const url3 = event.targetInfo.url;
-        const worker = new Worker(this._page, url3);
+        const worker = new Worker(this._page, url3, void 0, event.targetInfo.targetId, session2);
         this._page.addWorker(event.sessionId, worker);
         this._workerSessions.set(event.sessionId, session2);
         session2.once("Runtime.executionContextCreated", async (event2) => {
@@ -37834,7 +37996,9 @@
           session2.on("Inspector.workerScriptLoaded", () => worker.workerScriptLoaded());
         else
           worker.workerScriptLoaded();
-        session2._sendMayFail("Runtime.enable");
+        if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] === "0") {
+          session2._sendMayFail("Runtime.enable");
+        }
         this._crPage._networkManager.addSession(session2, this._page.frameManager.frame(event.targetInfo.parentFrameId ?? this._targetId) ?? void 0).catch(() => {
         });
         session2._sendMayFail("Runtime.runIfWaitingForDebugger");
@@ -38228,8 +38392,9 @@
           const message = new ConsoleMessage(null, this, event.type, void 0, args, stackTraceToLocation(event.stackTrace), event.timestamp);
           this.browserContext.emit(BrowserContext.Events.Console, message);
         });
-        session2.send("Runtime.enable", {}).catch((e) => {
-        });
+        if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] === "0")
+          session2.send("Runtime.enable", {}).catch((e) => {
+          });
         session2.send("Runtime.runIfWaitingForDebugger").catch((e) => {
         });
         session2.on("Inspector.targetReloadedAfterCrash", () => {
@@ -42876,7 +43041,11 @@
           }).catch((e) => null);
         });
         Promise.all([
-          session2.send("Runtime.enable"),
+          (() => {
+            if (process.env["REBROWSER_PATCHES_RUNTIME_FIX_MODE"] === "0") {
+              return session2.send("Runtime.enable", {});
+            }
+          })(),
           session2.send("Runtime.addBinding", { name: kBindingName2 }),
           session2.send("Page.enable"),
           session2.send("Page.addScriptToEvaluateOnNewDocument", { source: `
